### ISSUE TITLE
Getting joystick buttons state not working

### ISSUE THREAD

**Issue by j987456** [WhiteMagic/JoystickGremlin#272]
Dear WhiteMagic,
I started using Joystick Gremlin a few days ago as a replacement for FreePIE, so I'm mainly using scripts (plugins). The reason is that FreePIE is very CPU hungry. Running the same code FreePIE uses 3-5% CPU or even more, Joystick Gremlin only 0-1%. You did a very good job!
I was unable to get the _@gremlin.input_devices.periodic(\<seconds\>)_ decorator to work, then I saw that in the latest version of the program it is broken. However I'm also not able to get the state of buttons using _state = joy[device_guid].button(button_id).is_pressed_. For example this code gives me an error:
```
@joy_example.button(7)
def switch_OFF(event, joy, vjoy):
	
	# Event works correctly
	if event.is_pressed == True:
		vjoy[1].button(10).is_pressed = False
	else:
		vjoy[1].button(10).is_pressed = True

	# Getting the state of an other button gives an error
	state = joy["{5504V0B0-D77A-15E9-8001-123553541234}"].button(1).is_pressed
```
I'm doing something wrong?

Basically what I need is to set the vJoy buttons initial state, depending on how the joystick buttons are set, so that they are synchronized.
Without the period function I tried using a button event.

Any suggestions?
Thank you!

**Comment by WhiteMagic** [comment]
From a quick look at the code I suspect you have to provide a real GUID object to the joy[...] call. Converting the string to the object can be done using:
```
import gremlin.util
guid = gremlin.util.parse_guid("{GUID STRING HERE}")
```
Though even if this is fixed the end result might not be what you want. From what I gather you want to ensure your vJoy and joystick states are synchronized. The problem with that is that with DirectInput devices there is no way of getting the current state as you only get events when something changes. Gremlin will cache this when it sees something change but if a button or axis hasn't moved since Gremlin started it won't know what the state is.

**Comment by j987456** [comment]
Thank you for the answer, but I get an error using that code. I managed to get the buttons state using the code provided by muchimi here [https://github.com/WhiteMagic/JoystickGremlin/issues/246](https://github.com/WhiteMagic/JoystickGremlin/issues/246). This is the code:
```
j = joy[joystick_decorator.device_guid]
state = j.button(1).is_pressed
```

**EDIT
You can skip the part below and go directly to the next comment.**

This is the situation and the tests I did.
There are some joysticks that have switches with a lever with two positions. USB controllers often work like this: if the lever is ON, a button is held down, otherwise the button is simply released. However often in simulators it is necessary to be able to assign a button for the ON command and one for the OFF command (for example landing lights ON and landing lights OFF), it is not enough to release the ON command. Then you can use Joystick Gremlin so that it holds down a virtual vJoy button when the switch lever is in the OFF position and that releases it when the lever is ON. This way you can assign the two buttons in the simulators and everything works correctly.
However when Joystick Gremlin is activated all vJoy buttons are released so you have to synchronize those that must be kept pressed because the corresponding joystick lever is in the OFF position at that moment. I tried using a simple periodic function which is actually only executed once at startup.
This is the complete code:
```
import gremlin

joy_example = gremlin.input_devices.JoystickDecorator(
	"Joystick name",
	"{GUID}",
	"Default"
)

_buttons_synced = False
@gremlin.input_devices.periodic(1)
def sync_buttons(joy, vjoy):
	global _buttons_synced
	if _buttons_synced == False:
		j = joy[joy_example.device_guid]
		if not j.button(1).is_pressed:
			vjoy[1].button(1).is_pressed = True
		if not j.button(2).is_pressed:
			vjoy[1].button(2).is_pressed = True
		if not j.button(3).is_pressed:
			vjoy[1].button(3).is_pressed = True
		_buttons_synced = True

@joy_example.button(1)
def button_1_OFF(event, joy, vjoy):
	if event.is_pressed == True:
		vjoy[1].button(1).is_pressed = False
	else:
		vjoy[1].button(1).is_pressed = True

@joy_example.button(2)
def button_2_OFF(event, joy, vjoy):
	if event.is_pressed == True:
		vjoy[1].button(2).is_pressed = False
	else:
		vjoy[1].button(2).is_pressed = True

@joy_example.button(3)
def button_3_OFF(event, joy, vjoy):
	if event.is_pressed == True:
		vjoy[1].button(3).is_pressed = False
	else:
		vjoy[1].button(3).is_pressed = True
```
The strange thing is that the code works, but only if I activate Joystick Gremlin two times.
Let me explain. Starting with Joystick Gremlin closed, I open it and activate it: the virtual buttons are NOT synchronized. If I deactivate Joystick Gremlin and then activate it again a second time, without touching any button on the physical joystick, then it works, all virtual buttons are synchronized.
I hope I have been clear and that the explanation is useful!

**Comment by j987456** [comment]
Dear WhiteMagic,
after further tests I can confirm that the periodic functions work, but only if you activate Joystick Gremlin twice. You can see it by trying this simple code:
```
import gremlin

@gremlin.input_devices.periodic(0.1)
def periodic_function(vjoy):
	vjoy[1].button(1).is_pressed = True
```
When you start Joystick Gremlin and you activate it, the periodic function does not work.
If you deactivate Joystick Gremlin and then activate it again a second time, it works.

**Comment by WhiteMagic** [comment]
That's rather odd behaviour from that decorator, though it probably will make my life much easier in tracking down why it's not working the first time you activate a profile. Thanks a lot for tracking that detail down.

**Comment by j987456** [comment]
I'm glad I was helpful, thanks to you for this great app.

**Comment by WhiteMagic** [comment]
These issues with the periodic decorator should be resolved. If not feel free to reopen/recreate.
